src/man/man1/mbm_dslnavfix.1
.TH mbm_dslnavfix 1 "3 June 2013" "MB-System 5.0" "MB-System 5.0"
.SH NAME
\fBmbm_dslnavfix\fP \- Translate WHOI DSL AMS-120 navigation from
UTM eastings and northings to longitude and latitude.

.SH VERSION
Version 5.0

.SH SYNOPSIS
\fBmbm_dslnavfix\fP \fB\-F\fP\fInavformat\fP
\fB\-I\fP\fIinfile\fP \fB\-J\fP\fIutm_zone\fP
\fB\-O\fP\fIoutfile\fP
[\fB\-D\fP\fIdecimate\fP \fB\-E\fP\fIellipsoid\fP \fB\-H\fP \fB\-V\fP]

.SH DESCRIPTION
\fBmbm_dslnavfix\fP is a macro to translate
WHOI DSL AMS-120 navigation from
UTM eastings and northings to longitude and latitude.
The WHOI DSL group typically provides processed AMS-120
navigation is files separate from the bathymetry and
sidescan data. Moreover, the navigation is generally
sampled less frequently than the sonar
pings, and is reported in UTM easting and northing meters.
The \fBmbm_dslnavfix\fP macro is used to translate the
eastings and northings into the geographic coordinates
(longitude and latitude) used by MB-System programs. The
program \fBmbprocess\fP can be used to merge the translated
navigation in with the bathymetry and sidescan.

The DSL AMS-120 bathymetry and sidescan are provided in
either a set of parallel files supported as \fBMB-System\fP
format 111 (the old scheme) or in the OIC single file format
supported as \fBMB-System\fP format 141 (the new scheme).
In either case, we recommend translating the data to similar
but enhanced formats 112 and 142, respectively, to ease
processing.

We recommend the following procedure when
processing WHOI DSL AMS-120 data with MB-System:
.br
 	1) Concatenate all of the processed navigation
 	   files into a single file.
 		cat *.nav > cruise.rnav
 	2) Translate the DSL UTM navigation into geographic
 	   coordinates. In the "old" case the navigation is
           in a format designated by \fB\-F\fP\fI1\fP and
           in the new case by \fB\-F\fP\fI2\fP. Note that
           you must know the UTM zone used \- here we assume
           the zone is +27.
 		mbm_dslnavfix \-Icruise.rnav \\
 			-Ocruise.mrg2 \-F1 \-J27 \-V
 	3) Copy the bathymetry and sidescan data in each
 	   file from format 111 to format 112 or from
           format 141 to 142.
 		mbcopy \-F111/112 \\
 			-IDSL120.940630_1100.bat.dat \\
 			-ODSL120.940630_1100.mb112
 	4) Set the translated navigation to be merged with
           each of the format 112 or 142 files using
           \fBmbset\fP:
 		mbset \-I DSL120.940630_1100.mb112 \\
 			-P NAVFORMAT:2 \\
 			-P NAVFILE:cruise.mrg2 \\
 			-V
 	4) Process the data using \fBmbprocess\fP,
           producing a processed file called
           DSL120.940630_1100p.mb112 with longitude and
           latitude navigation.
 		mbprocess \-I DSL120.940630_1100.mb112

.SH AUTHORSHIP
David W. Caress (caress@mbari.org)
.br
  Monterey Bay Aquarium Research Institute
.br
Dale N. Chayes (dale@ldeo.columbia.edu)
.br
  Lamont-Doherty Earth Observatory

.SH OPTIONS
.TP
.B \-D
\fIdecimate\fP
.br
This option allows the user to decimate the input navigation.
The \fIdecimate\fP value represents the decimation factor.
The default is \fIdecimate\fP = 1, or no decimation.
.TP
.B \-E
\fIellipsoid\fP
.br
This option sets the ellipsoid used in the UTM projection.
The choices are:
 	WGS-84		1984 World Geodetic System
 	GRS-80		1980 International Geodetic Reference System
 	WGS-72		1972 World Geodetic System
 	Australian	1965 Used down under
 	Krasovsky	1940 Used in the Soviet Union
 	International	1924 Worldwide use
 	Hayford-1909	1909 Same as the International 1924
 	Clarke-1880	1880 Most of Africa, France
 	Clarke-1866	1866 North America, the Phillipines
 	Airy		1830 Great Britain
 	Bessel		1841 Central Europe, Chile, Indonesia
 	Everest		1830 India, Burma, Pakistan,  etc.
 	Sphere		1980 The mean radius in GRS-80
.br
Default: Clarke-1866 as per USGS standard.
.TP
.B \-F
\fInavformat\fP
.br
Sets the input navigation format. The available formats
include the old, pre-OIC format (\fB\-F\fP\fI1\fP), the new, OIC
associated formats (\fB\-F\fP\fI2\fP), and navigation obtained
using \fBmblist\fP with the \fB\-OtXY\fP option.
.TP
.B \-H
This "help" flag cause the program to print out a description
of its operation and then exit immediately.
.TP
.B \-I
\fIfilename\fP
.br
Sets the input WHOI DSL AMS-120 navigation filename.
.TP
.B \-J
\fIutm_zone\fP
.br
Sets the UTM zone used for the translation. The UTM zones
are 6 degrees wide in longitude and begin with zone 1 extending
from 180 W to 174 W, zone 2 extending from 174 W to 168 W, and
so on through zone 60. Negative zone
values are used for the Southern Hemisphere.
.TP
.B \-O
\fIoutfile\fP
.br
Sets the output filename for the translated navigation.
.TP
.B \-V
Causes \fBmbm_plot\fP to operate in "verbose" mode
so that it outputs
more information than usual.

.SH EXAMPLES
Suppose we have obtained a WHOI DSL navigation file
as outlined above and execute the macro as suggested:
.br
 		mbm_dslnavfix \-Icruise.rnav \\
 			-Ocruise.mrg2 \-F1 \-J27 \-V
.br
The output will appear as follows:

 Running mbm_dslnavfix...

 Program Status:
 \--------------

   Input DSL navigation format:   1
   Input DSL navigation file:     dsl.nav
   Output navigation file:        dsl.mrg2
   Temporary projected nav file:  tmp_proj_8044.nav
   Temporary geographic nav file: tmp_geo_8044.nav

   \------------

   Projection:                    UTM Zone 27
   Projection units:              meters

   \------------

 586 navigation records read...

 Running mapproject...
 mapproject tmp_proj_8044.nav \-Ju27/1:1000000 \-I \\
 	-R-20:0/0:0/-19:0/1:0r > tmp_geo_8044.nav

 Deleting tmp_proj_8044.nav, tmp_geo_8044.nav...

 Done...

.SH BUGS
Let us know...


NEWS
-------------------------------------------------------------------------------
MB-SYSTEM NEWS FILE:
-------------------------------------------------------------------------------
Version: $Id$
-------------------------------------------------------------------------------

In the GNU package standard, the NEWS file lists the history of user-visible
changes to a package. For the the MB-System open source software package for the
processing and display of swath sonar data, this information is included in the
ChangeLog file. Both the NEWS file and the ChangeLog file are located at the top
of the MB-System source code distribution directory structure.

-------------------------------------------------------------------------------


branches/csiro-trunk/src/macros/mbm_dslnavfix.pl
eval '(exit $?0)' && eval 'exec perl -S $0 ${1+"$@"}'
                    & eval 'exec perl -S $0 $argv:q'
                         if 0;
#--------------------------------------------------------------------
#    The MB-system:	mbm_dslnavfix.perl	8/9/96
#    $Id$
#
#    Copyright (c) 1996-2012 by
#    D. W. Caress (caress@mbari.org)
#      Monterey Bay Aquarium Research Institute
#      Moss Landing, CA
#    and D. N. Chayes (dale@lamont.ldgo.columbia.edu)
#      Lamont-Doherty Earth Observatory
#      Palisades, NY  10964
#
#    See README file for copying and redistribution conditions.
#--------------------------------------------------------------------
#
# Command:
#   mbm_dslnavfix
#
# Purpose:
#   Macro to take WHOI DSL AMS-120 processed navigation
#   in UTM projected eastings and northings and produce
#   navigation in longitude and latitude. The output navigation
#   can be merged with the DSL AMS-120 bathymetry and sidescan
#   data using mbmerge (use the -M2 option of mbmerge).
#
# Basic Usage:
#   mbm_dslnavfix -Iinfile -Ooutfile -Jutm_zone [-Ddecimate -V -H]
#
# Author:
#   David W. Caress
#   Lamont-Doherty Earth Observatory
#   Palisades, NY  10964
#   August 9, 1996
#
# Version:
#   $Id$
#
# Revisions:
#   $Log: mbm_dslnavfix.perl,v $
#   Revision 5.2  2003/04/17 20:42:48  caress
#   Release 5.0.beta30
#
#   Revision 5.1  2001/03/22 21:05:45  caress
#   Trying to make release 5.0.beta0
#
# Revision 5.0  2000/12/01  22:58:01  caress
# First cut at Version 5.0.
#
# Revision 4.6  2000/10/03  21:42:17  caress
# Snapshot for Dale.
#
# Revision 4.5  1999/12/29  00:17:55  caress
# Release 4.6.8
#
# Revision 4.4  1999/10/21  20:42:32  caress
# Added nav format to mbm_dslnavfix.
#
# Revision 4.3  1999/03/31  18:09:36  caress
# MB-System 4.6beta7
#
# Revision 4.2  1997/07/25  13:50:21  caress
# Added ellipsoid option.
#
# Revision 4.1  1997/04/21  16:54:41  caress
# MB-System 4.5 Beta Release.
#
# Revision 4.1  1997/03/26  15:27:16  caress
# Fixed handling of UTM zone as per Del Bohnenstiehl.
#
# Revision 4.0  1996/08/12  21:14:19  caress
# Initial revision.
#
#
#
$program_name = "mbm_dslnavfix";

# Deal with command line arguments
$command_line = "@ARGV";
&MBGetopts('D:d:E:e:F:f:HhI:i:J:j:O:o:Vv');
$decimate = 		($opt_D || $opt_d || 1);
$ellipsoid_use = 	($opt_E || $opt_e || "WGS-84");
$format =    		($opt_F || $opt_f || 2);
$help =    		($opt_H || $opt_h);
$infile =    		($opt_I || $opt_i);
$map_scale =    	($opt_J || $opt_j);
$outfile =    		($opt_O || $opt_o);
$verbose =    		($opt_V || $opt_v);

# print out help message if required
if ($help)
	{
	print "\n$program_name:\n";
	print "\nVersion: $Id$\n";
	print "\nMacro to take WHOI DSL AMS-120 processed navigation \n";
	print "in UTM projected eastings and northings and produce \n";
	print "navigation in longitude and latitude. The output navigation \n";
	print "can be merged with the DSL AMS-120 bathymetry and sidescan \n";
	print "data using mbmerge (use the -M2 option of mbmerge). \n";
        print "\nBasic Usage:  \n";
        print "mbm_dslnavfix -Iinfile -Ooutfile -Jutm_zone -F1 [-Ddecimate -V -H] \n\n";
	exit 0;
	}

# check for input and output files
if (!$infile)
	{
	print "\a";
	die "No input file specified - $program_name aborted\n";
	}
elsif (! -e $infile)
	{
	print "\a";
	die "Specified input file cannot be opened - $program_name aborted\n";
	}
if (!$outfile)
	{
	print "\a";
	die "No output file specified - $program_name aborted\n";
	}

# check for projection
if (!$map_scale)
	{
	print "\a";
	die "No UTM zone specified - $program_name aborted\n";
	}

# tell the world we got started
if ($verbose)
	{
	print "\nRunning $program_name...\n";
	}

# make the temporary filenames
$tmp_proj_navfile = "tmp_proj_$$.nav";
$tmp_geo_navfile = "tmp_geo_$$.nav";

# get projection
$projection = "u" . "$map_scale" . "/1:1000000";
$projection_id = "UTM Zone $map_scale";

# set units of eastings and northings
$projection_units = "M";
$units_meters = 1;

# now set the other projection parameters
if ($map_scale > 0)
	{
	$org_lon_n = -183 + 6 * $map_scale;
	}
else
	{
	$org_lon_n = -183 - 6 * $map_scale;
	}
$org_lon = "$org_lon_n:0";
$org_lat = "0:0";
$org_lon_n = $org_lon_n + 1;
$org_lon_plus1 = "$org_lon_n:0";
$org_lat_plus1 = "1:0";
if ($map_scale > 0)
	{
	$org_x_ft = 1640416.66667;
	$org_x_m = 500000.0;
	$org_y_ft = 0.0;
	$org_y_m = 0.0;
	}
else
	{
	$org_x_ft = 1640416.66667;
	$org_x_m = 500000.0;
	$org_y_ft = 32808399.0;
	$org_y_m = 100000000.0;
	}

$bounds = "$org_lon" . "/" . "$org_lat"
		. "/" . "$org_lon_plus1"
		. "/" . "$org_lat_plus1" . "r";

# set gmt defaults
($d_format) = `gmtdefaults -L | grep D_FORMAT` =~ /\S+\s+\S+\s+(\S+)/;
($ellipsoid_save) = `gmtdefaults -L | grep ELLIPSOID` =~ /\S+\s+\S+\s+(\S+)/;
`gmtset D_FORMAT %.10lg ELLIPSOID $ellipsoid_use`;

# print out info
if ($verbose)
	{
	print "\nProgram Status:\n";
	print "--------------\n\n";
	print "  Input DSL navigation format:   $format\n";
	print "  Input DSL navigation file:     $infile\n";
	print "  Output navigation file:        $outfile\n";
	print "  Temporary projected nav file:  $tmp_proj_navfile\n";
	print "  Temporary geographic nav file: $tmp_geo_navfile\n";
	if ($decimate)
		{
		print "  Decimation factor:             $decimate\n";
		}
	print "\n  ------------\n\n";
	print "  Projection:                    $projection_id\n";
	if ($units_feet == 1)
		{
		print "  Projection units:              feet\n";
		}
	else
		{
		print "  Projection units:              meters\n";
		}
	print "\n  ------------\n\n";
	}

# get number of lines
$linecountcom = "cat $infile | wc -l";
$numberlines = `$linecountcom`;

# now open the raw DSL navigation data
if (!open(INP,"<$infile"))
	{
	die "\nInput file $infile cannot be opened!\n$program_name aborted\n";
	}

# now open the tmp projected navigation file
if (!open(OTMP,">$tmp_proj_navfile"))
	{
	die "\nTmp file $tmp_proj_navfile cannot be opened!\n$program_name aborted\n";
	}

# now read the raw DSL navigation data
$daysec_old = -9999;
$count = $decimate - 1;
$lastcount = $count + $numberlines - 1;
while ($line = <INP>)
	{
	if ($format == 1)
		{
		if (($date, $time, $easting, $northing)
			= $line =~ /^\S+\s+(\S+)\s+(\S+)\s+\S+\s+\S+\s+\S+\s+(\S+)\s+(\S+)/)
			{
			# get time stamp
			($year, $month, $day)
				= $date =~ /^(.{2}).{1}(.{2}).{1}(.{2})/;
			($hour, $minute, $second)
				= $time =~ /^(.{2}).{1}(.{2}).{1}(.{5})/;
			if ($year > 61)
				{
				$year = $year + 1900;
				}
			else
				{
				$year = $year + 2000;
				}
			$daysec = $second + 60 * ($minute + 60 * $hour);
			if ($daysec != $daysec_old)
				{
				$count++;
				}
			if ($daysec != $daysec_old &&
				($count >= $decimate || $count == $lastcount))
				{
				$count = 0;

				push(@nyear, $year);
				push(@nmonth, $month);
				push(@nday, $day);
				push(@nhour, $hour);
				push(@nminute, $minute);
				push(@nsecond, $second);

				# get projected values in desired units
				if ($units_feet)
					{
					$xx = 12.0 / 1000000 *($easting - $org_x_ft);
					$yy = 12.0 / 1000000 *($northing - $org_y_ft);
					}
				elsif ($units_meters)
					{
					$xx = ($easting - $org_x_m) / (0.0254 * 1000000);
					$yy = ($northing - $org_y_m) / (0.0254 * 1000000);
					}

				#print "$year $month $day $hour:$minute:$second $easting $northing\n";
				print OTMP "$xx $yy\n";
				}
			$daysec_old = $daysec;
			}
		else
			{
			print "Unable to parse navigation from line:\n$line\n";
			}
		}
	elsif ($format == 2)
		{
		if (($date, $time, $easting, $northing)
			= $line =~ /^(\S+),(\S+),\S+,\S+,\S+,(\S+),(\S+)/)
			{
			# get time stamp
			($year, $month, $day)
				= $date =~ /^(.{2}).{1}(.{2}).{1}(.{2})/;
			($hour, $minute, $second)
				= $time =~ /^(.{2}).{1}(.{2}).{1}(.{2})/;
			if ($year > 61)
				{
				$year = $year + 1900;
				}
			else
				{
				$year = $year + 2000;
				}
			$daysec = $second + 60 * ($minute + 60 * $hour);
			if ($daysec != $daysec_old)
				{
				$count++;
				}
			if ($daysec != $daysec_old &&
				($count >= $decimate || $count == $lastcount))
				{
				$count = 0;

				push(@nyear, $year);
				push(@nmonth, $month);
				push(@nday, $day);
				push(@nhour, $hour);
				push(@nminute, $minute);
				push(@nsecond, $second);

				# get projected values in desired units
				if ($units_feet)
					{
					$xx = 12.0 / 1000000 *($easting - $org_x_ft);
					$yy = 12.0 / 1000000 *($northing - $org_y_ft);
					}
				elsif ($units_meters)
					{
					$xx = ($easting - $org_x_m) / (0.0254 * 1000000);
					$yy = ($northing - $org_y_m) / (0.0254 * 1000000);
					}

				#print "$year $month $day $hour:$minute:$second $easting $northing\n";
				print OTMP "$xx $yy\n";
				}
			$daysec_old = $daysec;
			}
		else
			{
			print "Unable to parse navigation from line:\n$line\n";
			}
		}
	elsif ($format == 3)
		{
		if (($date, $time, $easting, $northing)
			= $line =~ /^(\S+)-(\S+) \S+ \S+ \S+ (\S+) (\S+)/)
			{
			# get time stamp
			($year, $month, $day)
				= $date =~ /^(.{2}).{1}(.{2}).{1}(.{2})/;
			($hour, $minute, $second)
				= $time =~ /^(.{2}).{1}(.{2}).{1}(.{2})/;
			if ($year > 61)
				{
				$year = $year + 1900;
				}
			else
				{
				$year = $year + 2000;
				}
			$daysec = $second + 60 * ($minute + 60 * $hour);
			if ($daysec != $daysec_old)
				{
				$count++;
				}
			if ($daysec != $daysec_old &&
				($count >= $decimate || $count == $lastcount))
				{
				$count = 0;

				push(@nyear, $year);
				push(@nmonth, $month);
				push(@nday, $day);
				push(@nhour, $hour);
				push(@nminute, $minute);
				push(@nsecond, $second);

				# get projected values in desired units
				if ($units_feet)
					{
					$xx = 12.0 / 1000000 *($easting - $org_x_ft);
					$yy = 12.0 / 1000000 *($northing - $org_y_ft);
					}
				elsif ($units_meters)
					{
					$xx = ($easting - $org_x_m) / (0.0254 * 1000000);
					$yy = ($northing - $org_y_m) / (0.0254 * 1000000);
					}

				#print "$year $month $day $hour:$minute:$second $easting $northing\n";
				print OTMP "$xx $yy\n";
				}
			$daysec_old = $daysec;
			}
		else
			{
			print "Unable to parse navigation from line:\n$line\n";
			}
		}
	elsif ($format == 4)
		{
		if (($year, $month, $day, $hour, $minute, $second, $easting, $northing)
			= $line =~ /^(\S+)\s+(\S+)\s+(\S+)\s+(\S+)\s+(\S+)\s+(\S+)\s+(\S+)\s+(\S+)/)
			{
			$daysec = $second + 60 * ($minute + 60 * $hour);
			if ($daysec != $daysec_old)
				{
				$count++;
				}
			if ($daysec != $daysec_old &&
				($count >= $decimate || $count == $lastcount))
				{
				$count = 0;

				push(@nyear, $year);
				push(@nmonth, $month);
				push(@nday, $day);
				push(@nhour, $hour);
				push(@nminute, $minute);
				push(@nsecond, $second);

				# get projected values in desired units
				if ($units_feet)
					{
					$xx = 12.0 / 1000000 *($easting - $org_x_ft);
					$yy = 12.0 / 1000000 *($northing - $org_y_ft);
					}
				elsif ($units_meters)
					{
					$xx = ($easting - $org_x_m) / (0.0254 * 1000000);
					$yy = ($northing - $org_y_m) / (0.0254 * 1000000);
					}

				#print "$year $month $day $hour:$minute:$second $easting $northing\n";
				print OTMP "$xx $yy\n";
				}
			$daysec_old = $daysec;
			}
		else
			{
			print "Unable to parse navigation from line:\n$line\n";
			}
		}
	}

# close files
close INP;
close OTMP;

# get number of time stamps
$ntime = @nyear;

# print out info
if ($verbose)
	{
	print "\n$ntime navigation records read...\n";
	}

# unproject the eastings and northings into longitude and latitude
if ($verbose)
	{
	print "\nRunning mapproject...\n";
	print "mapproject $tmp_proj_navfile -J$projection -I -R$bounds > $tmp_geo_navfile\n";
	}
@line = `mapproject $tmp_proj_navfile -J$projection -I -R$bounds > $tmp_geo_navfile`;
if ($verbose)
	{
	foreach $line (@line) {
		print "$line";
		}
	}

# reset the gmt defaults
`gmtset D_FORMAT $d_format ELLIPSOID $ellipsoid_save`;

# now open the unprojected navigation data
if (!open(INP,"<$tmp_geo_navfile"))
	{
	die "\nTmp file $tmp_geo_navfile cannot be opened!\n$program_name aborted\n";
	}

# now read the unprojected navigation data
while ($line = <INP>)
	{
	if (($lon, $lat)
		= $line =~ /^(\S+)\s+(\S+)/)
		{
		# put lon in -180 to 180 range
		if ($lon < -180.0)
			{
			$lon = $lon + 360.0;
			}
		elsif ($lon > 180.0)
			{
			$lon = $lon - 360.0;
			}

		# add to list
		push(@nlon, $lon);
		push(@nlat, $lat);
		}
	else
		{
		die "\nUnable to parse navigation from line:\n$line\n$program_name aborted\n";
		}
	}

# close file
close INP;

# get number of unprojected nav points
$nnav = @nlon;

# check that numbers of data match
if ($ntime != $nnav)
	{
	die "\nNumber of time stamps ($ntime) disagrees with number of nav points ($nnav)!\n$program_name aborted\n";
	}

# now open the final navigation file
if (!open(OTMP,">$outfile"))
	{
	die "\nTmp file $outfile cannot be opened!\n$program_name aborted\n";
	}

# now write the final navigation to the file
for ($i = 0; $i < $ntime; $i ++)
	{
	print OTMP "$nyear[$i] $nmonth[$i] $nday[$i] $nhour[$i] $nminute[$i] $nsecond[$i] $nlon[$i] $nlat[$i]\n";
	}

# close file
close OTMP;

# remove the temporary files
if ($verbose)
	{
	print "\nDeleting $tmp_proj_navfile, $tmp_geo_navfile...\n";
	}
`rm -f $tmp_proj_navfile $tmp_geo_navfile`;

# done
if ($verbose)
	{
	print "\nDone...\n";
	}

exit 0;


#-----------------------------------------------------------------------
# This version of Getopts has been augmented to support multiple
# calls to the same option. If an arg in argumentative is followed
# by "+" rather than ":",  then the corresponding scalar will
# be concatenated rather than overwritten by multiple calls to
# the same arg.
#
# Usage:
#      do Getopts('a:b+c'); # -a takes arg, -b concatenates args,
#			    # -c does not take arg. Sets opt_* as a
#                           # side effect.

sub MBGetopts {
    local($argumentative) = @_;
    local(@args,$_,$first,$rest);
    local($errs) = 0;

    @args = split( / */, $argumentative );
    while(@ARGV && ($_ = $ARGV[0]) =~ /^-(.)(.*)/) {
	($first,$rest) = ($1,$2);
	$pos = index($argumentative,$first);
	if($pos >= $[) {
	    if($args[$pos+1] eq ':') {
		shift(@ARGV);
		if($rest eq '') {
		    ++$errs unless @ARGV;
		    $rest = shift(@ARGV);
		}
		eval "\$opt_$first = \$rest;";
		eval "\$flg_$first = 1;";
	    }
	    elsif($args[$pos+1] eq '+') {
		shift(@ARGV);
		if($rest eq '') {
		    ++$errs unless @ARGV;
		    $rest = shift(@ARGV);
		}
		if (eval "\$opt_$first") {
		    eval "\$opt_$first = \$opt_$first
				. \":\" . \$rest;";
		}
		else {
		    eval "\$opt_$first = \$rest;";
		}
		eval "\$flg_$first = 1;";
	    }
	    elsif($args[$pos+1] eq '%') {
		shift(@ARGV);
		if($rest ne '') {
		    eval "\$opt_$first = \$rest;";
		}
		else {
		    $rest = $ARGV[0];
		    ($one) = $rest =~ /^-(.).*/;
		    $pos = index($argumentative,$one);
		    if(!$one || $pos < $[) {
			eval "\$opt_$first = \$rest;";
			shift(@ARGV);
		    }
		}
		eval "\$flg_$first = 1;";
	    }
	    else {
		eval "\$opt_$first = 1";
		eval "\$flg_$first = 1;";
		if($rest eq '') {
		    shift(@ARGV);
		}
		else {
		    $ARGV[0] = "-$rest";
		}
	    }
	}
	else {
	    print STDERR "Unknown option: $first\n";
	    ++$errs;
	    if($rest ne '') {
		$ARGV[0] = "-$rest";
	    }
	    else {
		shift(@ARGV);
	    }
	}
    }
    $errs == 0;
}
#-----------------------------------------------------------------------
